lib/features/build_server/compiler_interface.dart

```diff
@@ -0,0 +1,270 @@
+    _updateState(CompileStatusMessage(
+      status: CompileStatus.waiting,
+      payload: WaitReason.finishing,
+    ));
+  }
+
+  @override
+  void done() {
+    _updateState(CompileStatusMessage(
+      status: CompileStatus.done,
+      payload: null,
+    ));
+  }
+}
+
+/// The interface to an isolate which manages a single project.
+class MainInterface extends Interface {
+  /// Callback to be used when the compilation state updates.
+  void Function(CompileStatusMessage)? onStateUpdate;
+
+  /// Constrcutor which should not be called unless for testing.
+  /// Use [launch] to start the main interface.
+  /// Otherwise the corresponding worker interface will not be invoked
+  /// automatically.
+  @visibleForTesting
+  MainInterface(SpawnedIsolate isolate) : super(isolate);
+
+  /// Launches the [MainInterface] by spawning a worker isolate and setting up the connection.
+  static Future<MainInterface> launch() async {
+    return MainInterface(
+      await connect(ReceivePort(), WorkerInterface.start),
+    );
+  }
+
+  @override
+  Future<void> onMessage(InterfaceMessage message) async {
+    if (message.name == _MessageIdentifiers.stateUpdate && message.args is CompileStatusMessage) {
+      onStateUpdate?.call(message.args as CompileStatusMessage);
+    }
+  }
+
+  /// `true` when the worker can handle a build request.
+  Future<bool> hasCapacity() async => await expectResponse(
+        InterfaceMessage(_MessageIdentifiers.hasCapacity, null),
+        typeError: '"${_MessageIdentifiers.hasCapacity}" expects a boolean response.',
+        timeout: Duration(seconds: 2),
+      );
+
+  /// Initializes the worker with the given [config], [files] and [cache].
+  Future<void> init(
+    Map<String, String> files,
+    CompilerConfig config,
+    ProjectCache? cache,
+  ) async =>
+      await expectResponse(
+        InterfaceMessage(_MessageIdentifiers.init, [files, config, cache]),
+        timeout: Duration(seconds: 10),
+      );
+
+  /// Updates the project with the given [config] and [files].
+  Future<List<String>> sync(
+    Map<String, String> files,
+    CompilerConfig config,
+  ) async =>
+      await expectResponse(
+        InterfaceMessage(_MessageIdentifiers.sync, [files, config]),
+        timeout: Duration(seconds: 10),
+      );
+
+  /// Updates the project with the given [file].
+  Future<void> update(SourceFile file) async => await expectResponse(
+        InterfaceMessage(_MessageIdentifiers.update, await file.asMemoryData()),
+        timeout: Duration(seconds: 2),
+      );
+
+  /// Builds the project.
+  ///
+  /// Use [onStateUpdate] to get the compilation state.
+  Future<void> build() async => await call(InterfaceMessage(_MessageIdentifiers.build, null));
+
+  /// The latest available binary for the project.
+  Future<Uint8List?> get binary async {
+    return await expectResponse(
+      InterfaceMessage(
+        _MessageIdentifiers.getBinary,
+        null,
+      ),
+      timeout: Duration(seconds: 2),
```

Comment:
Are you sure 2sec is enough for bigger binarys? (Same goes for files in `update`)